plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoIcebergRestCatalogFactory.java

```diff
@@ -0,0 +1,100 @@
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import io.trino.hdfs.ConfigurationUtils;
+import io.trino.plugin.base.CatalogName;
+import io.trino.plugin.hive.NodeVersion;
+import io.trino.plugin.iceberg.IcebergConfig;
+import io.trino.plugin.iceberg.catalog.TrinoCatalog;
+import io.trino.plugin.iceberg.catalog.TrinoCatalogFactory;
+import io.trino.spi.security.ConnectorIdentity;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.rest.RESTSessionCatalog;
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+
+import javax.inject.Inject;
+
+import java.net.URI;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class TrinoIcebergRestCatalogFactory
+        implements TrinoCatalogFactory
+{
+    private static final Logger log = Logger.get(TrinoIcebergRestCatalogFactory.class);
+
+    private final CatalogName catalogName;
+    private final String trinoVersion;
+    private final URI serverUri;
+    private final Optional<String> credential;
+    private final Optional<String> token;
+    private final IcebergRestCatalogConfig.Security security;
+    private final boolean isUniqueTableLocation;
+
+    private volatile RESTSessionCatalog icebergCatalog;
+
+    @Inject
+    public TrinoIcebergRestCatalogFactory(
+            CatalogName catalogName,
+            IcebergRestCatalogConfig restConfig,
+            IcebergConfig icebergConfig,
+            NodeVersion nodeVersion)
+    {
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
+        this.trinoVersion = requireNonNull(nodeVersion, "nodeVersion is null").toString();
+        requireNonNull(restConfig, "restConfig is null");
+        this.serverUri = restConfig.getBaseUri();
+        this.credential = restConfig.getCredential();
+        this.token = restConfig.getToken();
+        this.security = restConfig.getSecurity();
+        requireNonNull(icebergConfig, "icebergConfig is null");
+        this.isUniqueTableLocation = icebergConfig.isUniqueTableLocation();
+    }
+
+    @Override
+    public TrinoCatalog create(ConnectorIdentity identity)
+    {
+        if (icebergCatalog == null) {
+            synchronized (this) {
+                if (icebergCatalog == null) {
+                    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+                    properties.put(CatalogProperties.URI, serverUri.toString());
+                    properties.put(CatalogProperties.FILE_IO_IMPL, "org.apache.iceberg.io.ResolvingFileIO");
+                    properties.put("trino-version", trinoVersion);
+                    if (security == IcebergRestCatalogConfig.Security.OAUTH2) {
+                        credential.ifPresent(v -> properties.put(OAuth2Properties.CREDENTIAL, v));
+                        token.ifPresent(v -> properties.put(OAuth2Properties.TOKEN, v));
+                    }
+
+                    icebergCatalog = new RESTSessionCatalog();
```

Comment:
The uninitialized instance of `RESTSessionCatalog` gets assigned to `icebergCatalog` and 
"escapes" the synchronization block.
Please don't use double-checked locking pattern. It's an unnecessary complexity and error-prone (as witness here)

Reply:
I thought the volatile field prevents that issue (this is one of the "compliant" examples I've seen), but regardless, I think you're right and I don't see that we need double check here.  Will remove it.

Reply:
Technically speaking, volatile estabilishes happens-before between the write to the field and the read.
In English, this means "write is immediate and all prior writes are flushed into memory". So the assigned object gets published immediately and used to be initialized later. 

The current code in the PR avoids that: (1) it's more robustly synchronized and (2) assigns object only after init is done.